File: tutorials/phase-2-backend/04-r-integration/src/examples/expressionAnalysis.js
/**
 * Expression Analysis Example
 * Demonstrates differential expression and correlation analysis
 */

import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { promises as fs } from 'fs';
import jsStats from '../services/jsStats.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const DATA_DIR = join(__dirname, '..', '..', 'data');

async function expressionAnalysisExample() {
  console.log(`
═══════════════════════════════════════════════════════════
  Expression Analysis Example
═══════════════════════════════════════════════════════════
`);

  // Load expression data (skip comment lines starting with #)
  const csvContent = await fs.readFile(join(DATA_DIR, 'expression_data.csv'), 'utf-8');
  const lines = csvContent.trim().split('\n').filter(line => line.trim() && !line.startsWith('#'));
  const headers = lines[0].split(',').map(h => h.trim());
  
  const data = lines.slice(1).map(line => {
    const values = line.split(',');
    const obj = {};
    headers.forEach((h, i) => {
      obj[h] = h === 'sample_id' || h === 'condition' 
        ? values[i]?.trim() 
        : parseFloat(values[i]);
    });
    return obj;
  });

  console.log('📊 Data Summary:');
  console.log('─'.repeat(60));
  console.log(`  Total samples: ${data.length}`);
  console.log(`  Tumor samples: ${data.filter(d => d.condition === 'tumor').length}`);
  console.log(`  Normal samples: ${data.filter(d => d.condition === 'normal').length}`);
  
  const genes = headers.filter(h => h !== 'sample_id' && h !== 'condition');
  console.log(`  Genes analyzed: ${genes.join(', ')}`);

  // Differential Expression Analysis
  console.log('\n📍 Differential Expression Analysis:');
  console.log('─'.repeat(60));
  
  const deResult = jsStats.differentialExpression(data);
  
  console.log('\n  Results (sorted by adjusted p-value):');
  console.log('  ┌──────────┬──────────┬──────────┬──────────┬──────────┬───────────┐');
  console.log('  │   Gene   │  Tumor   │  Normal  │  Log2FC  │  Adj.P   │   Status  │');
  console.log('  ├──────────┼──────────┼──────────┼──────────┼──────────┼───────────┤');
  
  for (const r of deResult.results) {
    const status = r.significant 
      ? (r.direction === 'up' ? '↑ UP' : '↓ DOWN')
      : '  NS';
    console.log(`  │ ${r.gene.padEnd(8)} │ ${r.tumor_mean.toFixed(2).padStart(8)} │ ${r.normal_mean.toFixed(2).padStart(8)} │ ${r.log2_fold_change.toFixed(2).padStart(8)} │ ${r.adjusted_pvalue.toFixed(4).padStart(8)} │ ${status.padStart(9)} │`);
  }
  console.log('  └──────────┴──────────┴──────────┴──────────┴──────────┴───────────┘');
  
  console.log(`\n  Summary:`);
  console.log(`    Total genes: ${deResult.n_genes}`);
  console.log(`    Significant: ${deResult.n_significant}`);
  console.log(`    Up-regulated: ${deResult.results.filter(r => r.significant && r.direction === 'up').length}`);
  console.log(`    Down-regulated: ${deResult.results.filter(r => r.significant && r.direction === 'down').length}`);

  // Volcano Plot Data
  console.log('\n📍 Volcano Plot Categories:');
  console.log('─'.repeat(60));
  
  const volcano = deResult.results.map(r => ({
    ...r,
    neg_log10_pvalue: -Math.log10(r.adjusted_pvalue),
    category: !r.significant ? 'not_significant' :
              r.log2_fold_change > 0.5 ? 'up_regulated' :
              r.log2_fold_change < -0.5 ? 'down_regulated' : 'not_significant'
  }));
  
  for (const r of volcano) {
    const marker = r.category === 'up_regulated' ? '🔴' :
                   r.category === 'down_regulated' ? '🔵' : '⚪';
    console.log(`  ${marker} ${r.gene}: log2FC=${r.log2_fold_change.toFixed(2)}, -log10(p)=${r.neg_log10_pvalue.toFixed(2)}`);
  }

  // Gene Correlation Analysis
  console.log('\n📍 Gene Correlation Analysis:');
  console.log('─'.repeat(60));
  
  const correlations = [];
  for (let i = 0; i < genes.length - 1; i++) {
    for (let j = i + 1; j < genes.length; j++) {
      const x = data.map(d => d[genes[i]]);
      const y = data.map(d => d[genes[j]]);
      correlations.push({
        gene1: genes[i],
        gene2: genes[j],
        correlation: jsStats.correlation(x, y)
      });
    }
  }
  
  // Sort by absolute correlation
  correlations.sort((a, b) => Math.abs(b.correlation) - Math.abs(a.correlation));
  
  console.log('\n  Top 10 Gene Correlations:');
  console.log('  ┌──────────┬──────────┬────────────┬──────────────────┐');
  console.log('  │  Gene 1  │  Gene 2  │    Corr    │   Relationship   │');
  console.log('  ├──────────┼──────────┼────────────┼──────────────────┤');
  
  for (const c of correlations.slice(0, 10)) {
    const rel = c.correlation > 0.7 ? 'Strong positive' :
                c.correlation > 0.3 ? 'Moderate positive' :
                c.correlation < -0.7 ? 'Strong negative' :
                c.correlation < -0.3 ? 'Moderate negative' : 'Weak';
    console.log(`  │ ${c.gene1.padEnd(8)} │ ${c.gene2.padEnd(8)} │ ${c.correlation.toFixed(3).padStart(10)} │ ${rel.padEnd(16)} │`);
  }
  console.log('  └──────────┴──────────┴────────────┴──────────────────┘');

  console.log(`
═══════════════════════════════════════════════════════════
  Analysis Complete
═══════════════════════════════════════════════════════════
`);
}

expressionAnalysisExample();

File: tutorials/phase-2-backend/04-r-integration/src/r-scripts/expression_analysis.R
# Differential Expression Analysis Script
# Compares gene expression between tumor and normal samples

# Load required libraries
if (!require("stats")) install.packages("stats", repos="http://cran.r-project.org")

# Function: T-test based differential expression
# Simple approach for tutorial purposes
differential_expression <- function(input_file, output_file = NULL) {
  data <- read.csv(input_file)
  
  # Get gene columns (exclude sample_id and condition)
  gene_cols <- setdiff(names(data), c("sample_id", "condition"))
  
  # Separate tumor and normal samples
  tumor <- data[data$condition == "tumor", ]
  normal <- data[data$condition == "normal", ]
  
  results <- data.frame(
    gene = character(),
    tumor_mean = numeric(),
    normal_mean = numeric(),
    log2_fold_change = numeric(),
    pvalue = numeric(),
    adjusted_pvalue = numeric(),
    stringsAsFactors = FALSE
  )
  
  for (gene in gene_cols) {
    tumor_expr <- tumor[[gene]]
    normal_expr <- normal[[gene]]
    
    # Perform t-test
    t_result <- t.test(tumor_expr, normal_expr)
    
    # Calculate fold change (already log2 data)
    tumor_mean <- mean(tumor_expr)
    normal_mean <- mean(normal_expr)
    log2fc <- tumor_mean - normal_mean
    
    results <- rbind(results, data.frame(
      gene = gene,
      tumor_mean = round(tumor_mean, 3),
      normal_mean = round(normal_mean, 3),
      log2_fold_change = round(log2fc, 3),
      pvalue = t_result$p.value,
      adjusted_pvalue = NA  # Will calculate after all tests
    ))
  }
  
  # Adjust p-values for multiple testing (Benjamini-Hochberg)
  results$adjusted_pvalue <- p.adjust(results$pvalue, method = "BH")
  
  # Add significance flags
  results$significant <- results$adjusted_pvalue < 0.05
  results$direction <- ifelse(results$log2_fold_change > 0, "up", "down")
  
  # Sort by adjusted p-value
  results <- results[order(results$adjusted_pvalue), ]
  
  result <- list(
    analysis_type = "differential_expression",
    n_tumor = nrow(tumor),
    n_normal = nrow(normal),
    n_genes = length(gene_cols),
    n_significant = sum(results$significant),
    results = results
  )
  
  if (!is.null(output_file)) {
    json_output <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE)
    writeLines(json_output, output_file)
  }
  
  return(result)
}

# Function: Volcano plot data preparation
volcano_data <- function(input_file, output_file = NULL) {
  result <- differential_expression(input_file)
  
  # Add -log10 p-value for volcano plot
  volcano <- result$results
  volcano$neg_log10_pvalue <- -log10(volcano$adjusted_pvalue)
  
  # Classify points for coloring
  volcano$category <- "not_significant"
  volcano$category[volcano$significant & volcano$log2_fold_change > 0.5] <- "up_regulated"
  volcano$category[volcano$significant & volcano$log2_fold_change < -0.5] <- "down_regulated"
  
  result <- list(
    analysis_type = "volcano_data",
    points = volcano,
    thresholds = list(
      pvalue = 0.05,
      log2fc = 0.5
    )
  )
  
  if (!is.null(output_file)) {
    json_output <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE)
    writeLines(json_output, output_file)
  }
  
  return(result)
}

# Function: Gene correlation analysis
gene_correlation <- function(input_file, output_file = NULL) {
  data <- read.csv(input_file)
  
  # Get gene columns
  gene_cols <- setdiff(names(data), c("sample_id", "condition"))
  gene_data <- data[, gene_cols]
  
  # Calculate correlation matrix
  cor_matrix <- cor(gene_data, method = "pearson")
  
  # Get significant correlations
  n <- nrow(data)
  t_stat <- cor_matrix * sqrt((n - 2) / (1 - cor_matrix^2))
  p_matrix <- 2 * pt(-abs(t_stat), df = n - 2)
  
  # Extract upper triangle
  correlations <- list()
  for (i in 1:(length(gene_cols) - 1)) {
    for (j in (i + 1):length(gene_cols)) {
      correlations[[length(correlations) + 1]] <- list(
        gene1 = gene_cols[i],
        gene2 = gene_cols[j],
        correlation = round(cor_matrix[i, j], 3),
        pvalue = p_matrix[i, j],
        significant = p_matrix[i, j] < 0.05
      )
    }
  }
  
  result <- list(
    analysis_type = "gene_correlation",
    n_samples = nrow(data),
    n_genes = length(gene_cols),
    correlation_matrix = as.list(as.data.frame(cor_matrix)),
    pairwise_correlations = correlations
  )
  
  if (!is.null(output_file)) {
    json_output <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE)
    writeLines(json_output, output_file)
  }
  
  return(result)
}

# Command line interface
args <- commandArgs(trailingOnly = TRUE)
if (length(args) >= 2) {
  analysis_type <- args[1]
  input_file <- args[2]
  output_file <- if (length(args) >= 3) args[3] else NULL
  
  if (analysis_type == "de") {
    result <- differential_expression(input_file, output_file)
  } else if (analysis_type == "volcano") {
    result <- volcano_data(input_file, output_file)
  } else if (analysis_type == "correlation") {
    result <- gene_correlation(input_file, output_file)
  }
  
  if (is.null(output_file)) {
    cat(jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE))
  }
}
